File: threads/src/main/java/im/threads/holders/ImageFromConsultViewHolder.java
package im.threads.holders;

import android.graphics.PorterDuff;
import android.support.annotation.DrawableRes;
import android.support.v4.content.ContextCompat;
import android.util.TypedValue;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageView;
import android.widget.RelativeLayout;
import android.widget.TextView;

import java.text.SimpleDateFormat;
import java.util.Date;

import im.threads.R;
import im.threads.model.ChatStyle;
import im.threads.model.FileDescription;
import im.threads.picasso_url_connection_only.Callback;
import im.threads.picasso_url_connection_only.Picasso;
import im.threads.utils.CircleTransform;
import im.threads.utils.FileUtils;
import im.threads.utils.MaskedTransformer;

/**
 * Created by yuri on 30.06.2016.
 */
public class ImageFromConsultViewHolder extends BaseHolder {
    private TextView mTimeStampTextView;
    private ImageView mImage;
    private ImageView mConsultAvatar;
    private SimpleDateFormat sdf = new SimpleDateFormat("HH:mm");
    private View filter;
    private View filterSecond;
    private ChatStyle style;

    public ImageFromConsultViewHolder(ViewGroup parent) {
        super(LayoutInflater.from(parent.getContext()).inflate(R.layout.item_image_from_consult, parent, false));
        mTimeStampTextView = (TextView) itemView.findViewById(R.id.timestamp);
        mImage = (ImageView) itemView.findViewById(R.id.image);
        mConsultAvatar = (ImageView) itemView.findViewById(R.id.consult_avatar);
        filter = itemView.findViewById(R.id.filter);
        filterSecond = itemView.findViewById(R.id.filter_second);
        style = ChatStyle.getInstance();

        filter.setBackgroundColor(ContextCompat.getColor(itemView.getContext(), style.chatHighlightingColor));
        filterSecond.setBackgroundColor(ContextCompat.getColor(itemView.getContext(), style.chatHighlightingColor));

        mConsultAvatar.getLayoutParams().height = (int) itemView.getContext().getResources().getDimension(style.operatorAvatarSize);
        mConsultAvatar.getLayoutParams().width = (int) itemView.getContext().getResources().getDimension(style.operatorAvatarSize);
        mTimeStampTextView.setTextColor(getColorInt(style.incomingImageTimeColor));
        mTimeStampTextView.getBackground().setColorFilter(getColorInt(style.incomingImageTimeBackgroundColor), PorterDuff.Mode.SRC_ATOP);
    }

    public void onBind(String avatarPath
            , FileDescription fileDescription
            , long timestamp
            , View.OnClickListener listener
            , View.OnLongClickListener longListener
            , boolean isDownloadError
            , boolean isChosen
            , boolean isAvatarVisible) {
        mTimeStampTextView.setOnClickListener(listener);
        mTimeStampTextView.setOnLongClickListener(longListener);
        mImage.setOnClickListener(listener);
        mImage.setOnLongClickListener(longListener);
        mConsultAvatar.setOnClickListener(listener);
        mConsultAvatar.setOnLongClickListener(longListener);
        filter.setOnClickListener(listener);
        filter.setOnLongClickListener(longListener);
        mTimeStampTextView.setText(sdf.format(new Date(timestamp)));
        mImage.setImageResource(0);
        if (fileDescription.getFilePath() != null && !isDownloadError) {
            Picasso.with(itemView.getContext())
                    .load(fileDescription.getFilePath())
                    .fit()
                    .centerCrop()
                    .transform(new MaskedTransformer(itemView.getContext(), MaskedTransformer.TYPE_CONSULT))
                    .into(mImage, new Callback() {
                        @Override
                        public void onSuccess() {

                        }

                        @Override
                        public void onError() {
                            mImage.setImageResource(style.imagePlaceholder);
                        }
                    });
        } else if (isDownloadError) {
            mImage.setImageResource(style.imagePlaceholder);
        }
        if (isChosen) {
            filter.setVisibility(View.VISIBLE);
            filterSecond.setVisibility(View.VISIBLE);
        } else {
            filter.setVisibility(View.INVISIBLE);
            filterSecond.setVisibility(View.INVISIBLE);
        }
        @DrawableRes int resId = style.defaultOperatorAvatar;
        if (isAvatarVisible) {
            float bubbleLeftMarginDp = itemView.getContext().getResources().getDimension(R.dimen.margin_quarter);
            int bubbleLeftMarginPx = ((int) TypedValue.applyDimension(TypedValue.COMPLEX_UNIT_DIP, bubbleLeftMarginDp, itemView.getResources().getDisplayMetrics()));
            RelativeLayout.LayoutParams lp = (RelativeLayout.LayoutParams) mImage.getLayoutParams();
            lp.setMargins(bubbleLeftMarginPx, lp.topMargin, lp.rightMargin, lp.bottomMargin);
            mImage.setLayoutParams(lp);

            mConsultAvatar.setVisibility(View.VISIBLE);
            if (avatarPath != null) {
                avatarPath = FileUtils.convertRelativeUrlToAbsolute(itemView.getContext(), avatarPath);
                Picasso.with(itemView.getContext())
                        .load(avatarPath)
                        .error(style.defaultOperatorAvatar)
                        .fit()
                        .transform(new CircleTransform())
                        .centerInside()
                        .noPlaceholder()
                        .into(mConsultAvatar);
            } else {
                Picasso.with(itemView.getContext())
                        .load(resId)
                        .fit()
                        .noPlaceholder()
                        .transform(new CircleTransform())
                        .centerInside()
                        .into(mConsultAvatar);
            }
        } else {
            mConsultAvatar.setVisibility(View.GONE);

            int avatarSizeRes = style.operatorAvatarSize;
            int avatarSizePx = itemView.getContext().getResources().getDimensionPixelSize(avatarSizeRes);
            int bubbleLeftMarginPx = itemView.getContext().getResources().getDimensionPixelSize(R.dimen.margin_half);
            int avatarLeftMarginPx = itemView.getContext().getResources().getDimensionPixelSize(R.dimen.margin_half);
            RelativeLayout.LayoutParams lp = (RelativeLayout.LayoutParams) mImage.getLayoutParams();
            lp.setMargins(avatarSizePx + bubbleLeftMarginPx + avatarLeftMarginPx, lp.topMargin, lp.rightMargin, lp.bottomMargin);
            mImage.setLayoutParams(lp);
        }
    }
}

File: src/main/java/im/threads/ui/holders/ImageFromConsultViewHolder.kt
package im.threads.ui.holders

import android.util.TypedValue
import android.view.LayoutInflater
import android.view.View
import android.view.View.OnLongClickListener
import android.view.ViewGroup
import android.widget.ImageView
import android.widget.LinearLayout
import android.widget.TextView
import androidx.appcompat.content.res.AppCompatResources
import androidx.core.content.ContextCompat
import androidx.core.graphics.BlendModeColorFilterCompat
import androidx.core.graphics.BlendModeCompat
import androidx.core.view.isVisible
import im.threads.R
import im.threads.business.imageLoading.ImageLoader
import im.threads.business.imageLoading.ImageModifications
import im.threads.business.models.ChatItem
import im.threads.business.models.ConsultPhrase
import im.threads.business.models.FileDescription
import im.threads.business.models.enums.AttachmentStateEnum
import im.threads.business.ogParser.OpenGraphParser
import im.threads.ui.utils.ColorsHelper
import im.threads.ui.utils.gone
import im.threads.ui.utils.invisible
import im.threads.ui.utils.visible
import im.threads.ui.widget.textView.BubbleTimeTextView
import io.reactivex.subjects.PublishSubject
import java.text.SimpleDateFormat
import java.util.Date
import java.util.Locale

class ImageFromConsultViewHolder(
    parent: ViewGroup,
    private val maskedTransformation: ImageModifications.MaskedModification?,
    highlightingStream: PublishSubject<ChatItem>,
    openGraphParser: OpenGraphParser
) : BaseImageHolder(
    LayoutInflater.from(parent.context).inflate(
        R.layout.ecc_item_image_from_consult,
        parent,
        false
    ),
    highlightingStream,
    openGraphParser,
    true
) {

    private val sdf = SimpleDateFormat("HH:mm", Locale.getDefault())

    private val errorTextView: TextView = itemView.findViewById(R.id.errorText)
    private val loaderLayout: LinearLayout = itemView.findViewById(R.id.loaderLayout)
    private val fileNameTextView: TextView = itemView.findViewById(R.id.fileName)
    private val loader: ImageView = itemView.findViewById(R.id.loader)
    private val loaderImage: ImageView = itemView.findViewById(R.id.loaderImage)
    private val imageLayoutWithSpace: LinearLayout = itemView.findViewById(R.id.imageLayoutWithSpace)
    private val timeStampLoader: BubbleTimeTextView = itemView.findViewById<BubbleTimeTextView>(R.id.timeStampLoader).apply {
        setTextColor(getColorInt(style.incomingImageTimeColor))
        if (style.incomingMessageTimeTextSize > 0) {
            setTextSize(
                TypedValue.COMPLEX_UNIT_PX,
                itemView.context.resources.getDimension(style.incomingMessageTimeTextSize)
            )
        }
        background.colorFilter = BlendModeColorFilterCompat.createBlendModeColorFilterCompat(
            getColorInt(style.incomingImageTimeBackgroundColor),
            BlendModeCompat.SRC_ATOP
        )
    }
    private val consultAvatar = itemView.findViewById<ImageView>(R.id.consultAvatar).apply {
        layoutParams.height = itemView.context.resources.getDimension(style.operatorAvatarSize).toInt()
        layoutParams.width = itemView.context.resources.getDimension(style.operatorAvatarSize).toInt()
    }

    init {
        itemView.findViewById<View>(R.id.delimiter)
            .setBackgroundColor(getColorInt(style.incomingDelimitersColor))
        setTextColorToViews(arrayOf(fileNameTextView), style.incomingMessageTextColor)
    }

    fun onBind(
        consultPhrase: ConsultPhrase,
        highlighted: Boolean,
        buttonClickListener: View.OnClickListener,
        onLongClickListener: OnLongClickListener,
        onAvatarClickListener: View.OnClickListener
    ) {
        subscribeForHighlighting(consultPhrase, itemView)
        applyBubbleLayoutStyle()
        ColorsHelper.setTextColor(errorTextView, style.errorMessageTextColor)
        timeStampTextView.setOnClickListener(buttonClickListener)
        timeStampTextView.setOnLongClickListener(onLongClickListener)
        timeStampTextView.text = sdf.format(Date(consultPhrase.timeStamp))
        timeStampLoader.setOnClickListener(buttonClickListener)
        timeStampLoader.setOnLongClickListener(onLongClickListener)
        timeStampLoader.text = sdf.format(Date(consultPhrase.timeStamp))
        rootLayout.setOnClickListener(buttonClickListener)
        rootLayout.setOnLongClickListener(onLongClickListener)
        image.setOnClickListener(buttonClickListener)
        image.setOnLongClickListener(onLongClickListener)
        showAvatar(consultAvatar, consultPhrase, onAvatarClickListener)
        consultPhrase.fileDescription?.let {
            when (it.state) {
                AttachmentStateEnum.PENDING -> showLoaderLayout(it)
                AttachmentStateEnum.ERROR -> showErrorLayout(it)
                else -> {
                    showCommonLayout(it)
                    moveTimeToImageLayout()
                }
            }
        } ?: run {
            image.setImageResource(0)
        }
        rootLayout.apply {
            setBackgroundColor(
                ContextCompat.getColor(
                    itemView.context,
                    if (highlighted) style.chatHighlightingColor else R.color.ecc_transparent
                )
            )
        }
    }

    private fun applyBubbleLayoutStyle() {
        loaderLayout.background = AppCompatResources.getDrawable(
            itemView.context,
            style.incomingMessageBubbleBackground
        )
        setPaddings(true, loaderLayout)
        setLayoutMargins(true, loaderLayout)
        loaderLayout.background.colorFilter =
            BlendModeColorFilterCompat.createBlendModeColorFilterCompat(
                getColorInt(style.incomingMessageBubbleColor),
                BlendModeCompat.SRC_ATOP
            )
    }

    private fun showLoadImageAnimation() {
        loaderImage.isVisible = true
        initAnimation(loaderImage, true)
        rotateAnim.start()
    }

    private fun stopLoadImageAnimation() {
        loaderImage.gone()
        rotateAnim.cancel()
    }

    private fun showLoaderLayout(fileDescription: FileDescription) {
        loaderLayout.visible()
        imageLayoutWithSpace.gone()
        timeStampLoader.visible()
        errorTextView.gone()
        fileNameTextView.text = fileDescription.incomingName
        initAnimation(loader, true)
    }

    private fun showErrorLayout(fileDescription: FileDescription) {
        loaderLayout.isVisible = true
        errorTextView.isVisible = true
        imageLayoutWithSpace.gone()
        timeStampLoader.visible()
        loader.setImageResource(getErrorImageResByErrorCode(fileDescription.errorCode))
        fileNameTextView.text = fileDescription.incomingName
        val errorString = getString(getErrorStringResByErrorCode(fileDescription.errorCode))
        errorTextView.text = errorString
        rotateAnim.cancel()
    }

    private fun showCommonLayout(fileDescription: FileDescription) {
        imageLayoutWithSpace.visible()
        loaderLayout.invisible()
        timeStampLoader.gone()
        errorTextView.gone()
        rotateAnim.cancel()
        val previewUri = getPreviewUri(fileDescription.getPreviewFileDescription())
        val fileUri = if (previewUri?.toString().isNullOrBlank()) {
            fileDescription.getPreviewFileDescription()?.downloadPath
        } else {
            previewUri?.toString()
        }
        val isStateReady = fileDescription.state == AttachmentStateEnum.READY
        if (isStateReady && !fileDescription.isDownloadError) {
            if (!fileUri.isNullOrEmpty()) {
                showLoadImageAnimation()
                ImageLoader.get()
                    .load(fileUri)
                    .autoRotateWithExif(true)
                    .errorDrawableResourceId(style.imagePlaceholder)
                    .scales(ImageView.ScaleType.FIT_XY, ImageView.ScaleType.CENTER_CROP)
                    .modifications(maskedTransformation)
                    .callback(object : ImageLoader.ImageLoaderCallback {
                        override fun onImageLoaded() {
                            stopLoadImageAnimation()
                        }

                        override fun onImageLoadError() {
                            stopLoadImageAnimation()
                        }
                    })
                    .into(image)
            } else {
                image.setImageResource(style.imagePlaceholder)
            }
        } else {
            stopLoadImageAnimation()
            image.setImageResource(style.imagePlaceholder)
        }
    }
}
